Kinvey-Xamarin/Realtime/Stream.cs

```diff
@@ -0,0 +1,292 @@
+// http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
+// software, you hereby accept such terms of service  (and any agreement referenced
+// therein) and agree that you have read, understand and agree to be bound by such
+// terms of service and are of legal age to agree to such terms with Kinvey.
+//
+// This software contains valuable confidential and proprietary information of
+// KINVEY, INC and is subject to applicable licensing agreements.
+// Unauthorized reproduction, transmission or distribution of this file and its
+// contents is a violation of applicable laws.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Kinvey
+{
+	/// <summary>
+	/// </summary>
+	public class Stream<T> where T : IStreamable
+	{
+		KinveyRealtimeDelegate<T> RealtimeCallback { get; set; }
+
+		KinveyRealtimeDelegate<string> routerCallback;
+
+		/// <summary>
+		/// Represents the name of the stream.
+		/// </summary>
+		public string StreamName { get; set; }
+
+		private AbstractClient KinveyClient { get; set; }
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="T:Kinvey.Stream`1"/> class.
+		/// </summary>
+		/// <param name="streamName">Name of the stream during stream creation.</param>
+		/// <param name="client">[optional] <see cref="AbstractClient"/> object, default is null.</param>
+		public Stream(string streamName, AbstractClient client = null)
+		{
+			StreamName = streamName;
+
+			KinveyClient = client ?? Client.SharedClient;
+		}
+
+		/// <summary>
+		/// Grant access to the stream for publishing and/or subscribing to specified users and/or user groups.
+		/// </summary>
+		/// <returns>True if the grant request was successful, otherwise false.</returns>
+		/// <param name="streamACL"><see cref="StreamAccessControlList"/> object used to determine which users should be granted access for publish and/or subscribe access.</param>
+		/// <param name="ct">[optional] CancellationToken used to cancel the request.</param>
+		public async Task<bool> GrantStreamAccess(StreamAccessControlList streamACL, CancellationToken ct = default(CancellationToken))
+		{
+			GrantStreamAccessRequest requestGrant = BuildGrantStreamAccessRequest(streamACL);
+			ct.ThrowIfCancellationRequested();
+			var success = await requestGrant.ExecuteAsync();
+			return (success != null);
+		}
+
+		/// <summary>
+		/// Publish a message of type {T} to the specified user.
+		/// </summary>
+		/// <param name="receiverID">Receiver identifier.</param>
+		/// <param name="message">Message.</param>
+		public async Task<bool> Publish(string receiverID, T message)
+		{
+			bool result = false;
+
+			// Make KCS request for publish access for the given receiverID
+			// KCS will return, if successful, a response which will include the PubNub channel name
+			JObject response = await RequestPublishAccess(receiverID);
```

Comment:
@vinaygahlawat This seems to be called on every message. That shouldn't be unnecessary.